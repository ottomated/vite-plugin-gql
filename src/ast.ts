import type { Node } from 'estree';
import { walk } from 'zimmerframe';

export function find_import(ast: Node, module_id: string): string | null {
	let import_name: string | null = null;
	walk(
		ast,
		{},
		{
			ImportDeclaration(node, { stop }) {
				if (node.source.value !== module_id) return;
				const specifier = node.specifiers.find(
					(s) => s.type === 'ImportDefaultSpecifier',
				);
				if (!specifier) return;
				import_name = specifier.local.name;
				stop();
			},
		},
	);
	return import_name;
}
