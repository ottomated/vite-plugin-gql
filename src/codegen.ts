import type {
	ASTNode,
	GraphQLSchema,
	GraphQLType,
	SelectionSetNode,
	SourceLocation,
	VariableDefinitionNode,
} from 'graphql';
import {
	GraphQLError,
	isEnumType,
	isInputObjectType,
	isInterfaceType,
	isListType,
	isNonNullType,
	isObjectType,
	isScalarType,
	isUnionType,
	parse,
	typeFromAST,
} from 'graphql';

export const SCALAR_TYPES = {
	Int: 'number',
	Float: 'number',
	String: 'string',
	Boolean: 'boolean',
	ID: 'string',
	URL: 'string',
	Decimal: 'number',
} as Record<string, string>;

export function generate_typescript(
	query: string,
	schema: GraphQLSchema,
	custom_scalars: Record<string, string> | undefined,
): {
	variables: string | null;
	return_type: string;
} {
	const ast = parse(query);
	if (ast.definitions.length > 1) {
		throw new GraphQLError(
			`Expected a single query, got ${ast.definitions.length}`,
			{
				nodes: [ast.definitions[1]!],
			},
		);
	}
	if (ast.definitions.length === 0) {
		throw new Error('Expected a query');
	}
	const definition = ast.definitions[0]!;
	if (definition.kind !== 'OperationDefinition') {
		throw new GraphQLError(`Expected an operation, got ${definition.kind}`, {
			nodes: [definition],
		});
	}
	const root = schema.getRootType(definition.operation);
	if (!root) {
		throw new GraphQLError(
			`Schema doesn't contain any ${definition.operation} operations`,
			{
				nodes: [definition],
			},
		);
	}
	const scalar_types = custom_scalars
		? { ...SCALAR_TYPES, ...custom_scalars }
		: SCALAR_TYPES;
	return {
		variables: generate_variable_types(
			definition.variableDefinitions,
			schema,
			scalar_types,
		),
		return_type: generate_selected_type(definition, root, schema, scalar_types),
	};
	// return generate_typedef(selection, operation.type, schema);
}

function generate_variable_types(
	variables: readonly VariableDefinitionNode[] | undefined,
	schema: GraphQLSchema,
	scalar_types: Record<string, string>,
): string | null {
	if (!variables || !variables.length) return null;
	return (
		'{' +
		variables.map((v) => {
			const type = typeFromAST(schema, v.type);
			const typescript = type
				? generate_selected_type(v.type, type, schema, scalar_types)
				: 'unknown';

			return `${v.variable.name.value}: ${typescript}`;
		}) +
		'}'
	);
}

function generate_selected_type(
	node: ASTNode & { selectionSet?: SelectionSetNode },
	type: GraphQLType,
	schema: GraphQLSchema,
	scalar_types: Record<string, string>,
): string {
	if (isListType(type)) {
		return `(Array<${generate_selected_type(node, type.ofType, schema, scalar_types)}>) | null | undefined`;
	}
	if (isNonNullType(type)) {
		const sub_type = generate_selected_type(
			node,
			type.ofType,
			schema,
			scalar_types,
		);
		if (is_null_type(sub_type)) {
			return sub_type.slice(1, -20);
		}
		return sub_type;
	}
	const selections = node.selectionSet?.selections ?? [];

	let typescript: string;

	if (isObjectType(type) || isInterfaceType(type)) {
		const fields = type.getFields();
		typescript = '{';
		for (const field of selections) {
			if (field.kind !== 'Field')
				throw new GraphQLError(`Expected a field, got ${field.kind}`, {
					nodes: [field],
				});
			const type_field = fields[field.name.value];
			if (!type_field) {
				throw new GraphQLError(
					`Property '${field.name.value}' does not exist on type ${type.name}`,
					{
						nodes: [field],
					},
				);
			}
			const typescript_type = generate_selected_type(
				field,
				type_field.type,
				schema,
				scalar_types,
			);
			typescript += `${JSON.stringify(field.name.value)}${is_null_type(typescript_type) ? '?' : ''}: ${typescript_type},`;
		}
		typescript += '}';
	} else if (isUnionType(type)) {
		const objects: Array<string> = [];
		for (const field of selections) {
			if (field.kind !== 'InlineFragment') {
				throw new GraphQLError(
					`Expected an inline fragment, got ${field.kind}`,
					{
						nodes: [field],
					},
				);
			}
			const union_member = type
				.getTypes()
				.find((t) => t.name === field.typeCondition?.name.value);
			if (!union_member) {
				throw new GraphQLError(
					`Union type '${type.name}' does not contain member '${field.typeCondition?.name.value}'`,
					{
						nodes: [field],
					},
				);
			}
			objects.push(
				`(${generate_selected_type(field, union_member, schema, scalar_types)})`,
			);
		}
		typescript = objects.join(' | ');
	} else if (isEnumType(type)) {
		typescript = type
			.getValues()
			.map((v) => JSON.stringify(v.name))
			.join(' | ');
	} else if (isScalarType(type)) {
		if (type.name in scalar_types) {
			typescript = scalar_types[type.name]!;
		} else {
			throw new GraphQLError(
				`Unknown scalar type '${type.name}' (add it to the 'customScalars' option in your vite config)`,
				{
					nodes: [node],
				},
			);
		}
	} else if (isInputObjectType(type)) {
		// Select all fields on an input object
		const fields = type.getFields();
		typescript = '{';
		for (const [name, field] of Object.entries(fields)) {
			const typescript_type = generate_selected_type(
				node,
				field.type,
				schema,
				scalar_types,
			);
			typescript += `${JSON.stringify(name)}${is_null_type(typescript_type) ? '?' : ''}: ${typescript_type},`;
		}
		typescript += '}';
	} else {
		throw new GraphQLError(`Unknown type '${type}'`, {
			nodes: [node],
		});
	}
	return `(${typescript}) | null | undefined`;
}

function is_null_type(type: string): boolean {
	return type.startsWith('(') && type.endsWith(') | null | undefined');
}

export function location_to_index(
	location: SourceLocation,
	code: string,
): number {
	const lines = code.split('\n');
	let index = 0;
	for (let i = 0; i < location.line - 1; i++) {
		index += lines[i]!.length + 1;
	}
	return index + location.column;
}
