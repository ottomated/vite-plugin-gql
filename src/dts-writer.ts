import { writeFile } from 'node:fs/promises';
import { basename } from 'node:path';
import { format } from 'prettier';

// You can undo this if you want just makes more consistent code
export type GqlTypeInfo =
	| {
			variables: string | null;
			return_type: string;
	  }
	| {
			error: string;
	  };

export class DtsWriter {
	out_file: string;
	module_id: string;

	#file_map: Map<string, Map<string, GqlTypeInfo>> = new Map();

	constructor(module_id: string, out_file: string) {
		this.module_id = module_id;
		this.out_file = out_file;
	}

	update_file(
		id: string,
		types: Map<string, GqlTypeInfo>,
	): void {
		this.#file_map.set(id, types);

		clearTimeout(this.write_timeout);
		this.write_timeout = setTimeout(() => {
			this.write_timeout = undefined;
			this.#write();
		}, 100);
	}

	write_timeout: ReturnType<typeof setTimeout> | undefined;

	async #write() {
		let dts = `// ==== This file is auto-generated by @o7/vite-plugin-gql. Do not edit. ====
		declare module '${this.module_id}' {\n`;

		for (const [file, types] of this.#file_map) {
			dts += `// ${basename(file)}\n`;
			for (const [query, typescript] of types) {
				if ('error' in typescript) {
					dts += `/** @deprecated ${typescript.error} */
					export default function gql(
						query: ${JSON.stringify(query)},
						variables?: any
					): Promise<{__gql_error: ${JSON.stringify(typescript.error)}}>;\n`;
					continue;
				}

				const { variables, return_type } = typescript;
				dts += `export default function gql(
					query: ${JSON.stringify(query)},
					variables${variables === null ? '?: undefined' : `: ${variables}`}
				): Promise<(${return_type})>;\n`;
			}
		}
		dts += '}';

		const formatted =
			'/* eslint-disable */\n/* prettier-ignore */\n' +
			(await format(dts, {
				parser: 'typescript',
				trailingComma: 'all',
				useTabs: true,
				singleQuote: true,
			}));

		await writeFile(this.out_file, formatted);
	}
}
