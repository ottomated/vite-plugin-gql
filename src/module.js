/* global GQL_URL, GQL_HEADERS */

export default async function gql(query, variables) {
	const variables_body = variables
		? `,"variables":${JSON.stringify(variables)}`
		: '';
	const response = await fetch(GQL_URL, {
		method: 'POST',
		headers: GQL_HEADERS,
		body: `{"query":${query}${variables_body}}`,
	});
	if (!response.ok) {
		throw new Error(
			`GQL error: ${response.statusText} ${await response.text()}`,
		);
	}
	return response.json();
}
