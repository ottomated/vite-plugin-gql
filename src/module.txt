/* global HEADERS */

export default async function gql(query, variables) {
	const response = await fetch(
		'https://mogul-merch-inc.myshopify.com/api/2025-01/graphql.json',
		{
			method: 'POST',
			headers: HEADERS,
			body: `{"query":${query},"variables":${JSON.stringify(variables)}}`,
		},
	);
	if (!response.ok) {
		throw new Error(
			`GQL error: ${response.statusText} ${await response.text()}`,
		);
	}
	const json = await response.json();
	if ('errors' in json) {
		throw new Error(
			'GQL error: ' + json.errors.map((e) => e.message).join(', '),
		);
	}
	return json.data;
}
