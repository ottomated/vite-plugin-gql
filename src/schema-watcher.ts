import { loadSchema, type LoadSchemaOptions } from '@graphql-tools/load';
import { UrlLoader } from '@graphql-tools/url-loader';
import type { GraphQLSchema } from 'graphql';
import type { WatchSchemaConfig } from './index';

export class SchemaWatcher {
	private config: WatchSchemaConfig;
	private schemaOptions?: Omit<LoadSchemaOptions, 'loaders' | 'headers'>;
	private onSchemaChange: (schema: GraphQLSchema) => void;
	private onError: (error: Error) => void;
	private currentSchema: GraphQLSchema | null = null;
	private intervalId: NodeJS.Timeout | null = null;

	constructor(
		config: WatchSchemaConfig,
		schemaOptions: Omit<LoadSchemaOptions, 'loaders' | 'headers'> | undefined,
		onSchemaChange: (schema: GraphQLSchema) => void,
		onError: (error: Error) => void
	) {
		this.config = config;
		this.schemaOptions = schemaOptions;
		this.onSchemaChange = onSchemaChange;
		this.onError = onError;
	}

	async startPolling(): Promise<void> {
		try {
			await this.loadAndUpdateSchema();
		} catch (error) {
			this.onError(error instanceof Error ? error : new Error(String(error)));
		}

		if (this.config.interval && this.config.interval > 0) {
			this.intervalId = setInterval(async () => {
				try {
					await this.loadAndUpdateSchema();
				} catch (error) {
					this.onError(error instanceof Error ? error : new Error(String(error)));
				}
			}, this.config.interval);
		}
	}

	stopPolling(): void {
		if (this.intervalId) {
			clearInterval(this.intervalId);
			this.intervalId = null;
		}
	}

	getCurrentSchema(): GraphQLSchema | null {
		return this.currentSchema;
	}

	private async loadAndUpdateSchema(): Promise<void> {
		const env = process.env;
		const url = typeof this.config.url === 'function' ? this.config.url(env) : this.config.url;
		
		let headers: Record<string, string> = {};
		if (this.config.headers) {
			if (typeof this.config.headers === 'function') {
				headers = this.config.headers(env);
			} else {
				for (const [key, value] of Object.entries(this.config.headers)) {
					if (typeof value === 'function') {
						headers[key] = value(env);
					} else {
						headers[key] = value;
					}
				}
			}
		}

		const controller = new AbortController();
		const timeoutId = this.config.timeout 
			? setTimeout(() => controller.abort(), this.config.timeout)
			: null;

		try {
			const schema = await loadSchema(url, {
				...this.schemaOptions,
				headers,
				loaders: [new UrlLoader()],
			});

			this.currentSchema = schema;
			this.onSchemaChange(schema);
		} finally {
			if (timeoutId) {
				clearTimeout(timeoutId);
			}
		}
	}
}