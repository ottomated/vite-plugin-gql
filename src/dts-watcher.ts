import { readFile, writeFile } from 'node:fs/promises';
import { basename, join, relative } from 'node:path';
import { format } from 'prettier';
import type { TypescriptDetails } from './codegen';
import { FSWatcher } from 'chokidar';
import type { GraphQLSchema } from 'graphql';
import type { ProgramNode } from 'rollup';
import { find_import, walk_ast } from './ast';
import { parseAst } from 'rollup/parseAst';

export type TypeMap = {
	[query: string]: TypescriptDetails | { error: string };
};

export class DtsWatcher {
	out_file: string;
	module_id: string;
	schema_promise: Promise<GraphQLSchema>;

	#custom_scalars: Record<string, string> | undefined;
	#file_map = new Map<string, TypeMap>();
	#watcher = new FSWatcher();

	constructor(
		module_id: string,
		out_file: string,
		schema_promise: Promise<GraphQLSchema>,
		custom_scalars: Record<string, string> | undefined,
	) {
		this.module_id = module_id;
		this.out_file = out_file;
		this.schema_promise = schema_promise;
		this.#custom_scalars = custom_scalars;
		this.#watcher.on('change', (path) => {
			this.#update_file_from_watch(path);
		});

		// read cached files
		readFile(this.out_file, 'utf-8')
			.then((text) => {
				const files_raw = text.match(/\/\/ <files>(.*)<\/files>/)?.[1];
				if (!files_raw) return;
				let files: string[];
				try {
					files = JSON.parse(files_raw);
				} catch {
					return;
				}
				if (!Array.isArray(files)) return;
				for (const file of files) {
					if (typeof file !== 'string') continue;
					this.#update_file_from_watch(join(process.cwd(), file));
				}
			})
			.catch(() => {});
	}

	update_file(id: string, types: TypeMap): void {
		const relative_path = relative(process.cwd(), id);

		this.#file_map.set(relative_path, types);
		this.#watcher.add(id);

		clearTimeout(this.write_timeout);
		this.write_timeout = setTimeout(() => {
			this.write_timeout = undefined;
			this.#write();
		}, 100);
	}

	write_timeout: ReturnType<typeof setTimeout> | undefined;

	async #update_file_from_watch(path: string) {
		const code = await readFile(path, 'utf-8').catch(() => null);
		if (code === null) {
			this.#watcher.unwatch(path);
			return;
		}
		let ast: ProgramNode;
		try {
			ast = parseAst(code);
		} catch (_) {
			return;
		}
		const import_name = find_import(ast, this.module_id);
		if (!import_name) return;

		const schema = await this.schema_promise;

		const types = walk_ast(
			{
				ast,
				schema,
				custom_scalars: this.#custom_scalars,
				import_name,
				throw_gql_errors: false,
				magic_string: undefined,
			},
			{
				// swallow warnings during typegen
				warn() {},
				error(err) {
					throw err;
				},
			},
		);

		this.update_file(path, types);
	}

	async #write() {
		let dts = `// ==== This file is auto-generated by @o7/vite-plugin-gql. Do not edit. ====
		// <files>${JSON.stringify(Array.from(this.#file_map.keys()))}</files>
		declare module '${this.module_id}' {
		export type GraphQLError = {
			message: string;
			locations?: Array<{
				line: number;
				column: number;
			}>;
			path?: Array<string | number>;
			extensions?: Record<string, unknown>;
		}\n`;

		for (const [file, types] of this.#file_map) {
			dts += `//#region ${basename(file)}\n`;
			for (const [query, typescript] of Object.entries(types)) {
				if ('error' in typescript) {
					dts += `/** @deprecated ${typescript.error} */
					export default function gql(
						query: ${JSON.stringify(query)},
						variables?: unknown
					): never & {Error: ${JSON.stringify(typescript.error)}};\n`;
					continue;
				}

				const { variables, return_type } = typescript;
				dts += `export default function gql(
					query: ${JSON.stringify(query)},
					variables${variables === null ? '?: undefined' : `: ${variables}`}
				): Promise<{data?: (${return_type}), errors?: Array<GraphQLError>}>;\n`;
			}
		}
		dts += '}';

		const formatted =
			'/* prettier-ignore */\n' +
			(await format(dts, {
				parser: 'typescript',
				trailingComma: 'all',
				useTabs: true,
				singleQuote: true,
			}));

		await writeFile(this.out_file, formatted);
	}
}
