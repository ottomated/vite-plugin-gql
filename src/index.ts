import type { Plugin } from 'vite';
import module from './module.txt';
import type { Expression, Node } from 'estree';
import type { ProgramNode } from 'rollup';
import { find_import } from './ast';
import MagicString from 'magic-string';
import { walk } from 'zimmerframe';
import { generate_typescript, location_to_index } from './codegen';
import { loadSchema, type LoadSchemaOptions } from '@graphql-tools/load';
import { UrlLoader } from '@graphql-tools/url-loader';
import type { GraphQLSchema } from 'graphql';
import { GraphQLError, stripIgnoredCharacters } from 'graphql';
import { DtsWriter } from './dts-writer';

type RollupNode<T> = T & { start: number; end: number };

interface PluginConfig {
	/**
	 * The place to import the gql tag from. Defaults to `$gql`.
	 */
	moduleId?: string;
	/**
	 * The URL to the GraphQL endpoint.
	 */
	url: string;
	/**
	 * Headers to send with the request.
	 */
	headers?: Record<string, string>;
	/**
	 * Options to pass to the schema loader.
	 */
	schemaOptions?: Omit<LoadSchemaOptions, 'loaders' | 'headers'>;
	/**
	 * The file to write the generated types to.
	 * @example "src/gql.d.ts"
	 */
	outFile: string;
	/**
	 * A map of GraphQL scalar types to their typescript types
	 * @example { ID: 'string' }
	 */
	customScalars?: Record<string, string>;
}

export default function gql_tag_plugin(config: PluginConfig): Plugin {
	const {
		moduleId = '$gql',
		url,
		customScalars,
		outFile,
		headers = {},
	} = config;
	if (moduleId.includes("'")) throw new Error('Invalid moduleId');

	if (!('Content-Type' in headers)) {
		headers['Content-Type'] = 'application/json';
	}

	let schema_promise: Promise<GraphQLSchema> | undefined;
	let is_build: boolean;

	const dts_writer = new DtsWriter(moduleId, outFile);

	return {
		name: 'vite-plugin-gql-tag',
		config(_, env) {
			is_build = env.command === 'build';
		},
		resolveId(id) {
			if (id === moduleId) {
				return '\0' + moduleId;
			}
		},
		load(id) {
			if (id === '\0' + moduleId) {
				return module
					.replace('URL', JSON.stringify(url))
					.replace('HEADERS', JSON.stringify(headers));
			}
		},
		async transform(code, id) {
			let ast: ProgramNode;
			try {
				ast = this.parse(code);
			} catch (_) {
				return;
			}
			const import_name = find_import(ast, moduleId);
			if (!import_name) return { code, ast };

			schema_promise ??= loadSchema(config.url, {
				...config.schemaOptions,
				headers,
				loaders: [new UrlLoader()],
			});
			const schema = await schema_promise;

			const s = new MagicString(code);
			const types = new Map<
				string,
				| {
						variables: string | null;
						return_type: string;
				  }
				| {
						error: string;
				  }
			>();
			const throw_error = this.error.bind(this);
			const throw_warning = this.warn.bind(this);
			walk(
				ast as RollupNode<Node>,
				{},
				{
					_(_, { next }) {
						next();
					},
					CallExpression(node, { next }) {
						if (node.callee.type !== 'Identifier') return next();
						if (node.callee.name !== import_name) return next();

						const query = node.arguments[0] as
							| RollupNode<Expression>
							| undefined;
						if (!query) {
							return throw_error(
								`${node.callee.name} requires a query argument`,
							);
						}
						let query_value: string;
						if (query.type === 'TemplateLiteral') {
							if (query.quasis.length !== 1) {
								return throw_error(
									`The query argument to ${node.callee.name} can't have interpolation`,
									query.start,
								);
							}
							query_value =
								query.quasis[0]!.value.cooked ?? query.quasis[0]!.value.raw;
						} else if (
							query.type === 'Literal' &&
							typeof query.value === 'string'
						) {
							query_value = query.value;
						} else {
							return throw_error(
								`The first argument to ${node.callee.name} must be a literal string (i.e. \`query { ... }\`)`,
								query.start,
							);
						}

						let minified: string;
						try {
							const typescript = generate_typescript(
								query_value,
								schema,
								customScalars,
							);
							types.set(query_value, typescript);
							// Double stringify - one to turn it into JS, one because it's
							// passed to a json api
							minified = JSON.stringify(
								JSON.stringify(stripIgnoredCharacters(query_value)),
							);
						} catch (e) {
							let location = query.start;
							let message = String(e);
							if (e instanceof GraphQLError && e.locations?.length) {
								location =
									location_to_index(e.locations[0]!, query_value) + query.start;
								message = e.message;
							}
							if (is_build) {
								return throw_error(message, location);
							} else {
								throw_warning(message, location);
							}
							types.set(query_value, { error: message });
							minified = `(() => { throw new Error(${JSON.stringify(
								message,
							)}); })()`;
						}

						s.update(query.start, query.end, minified);
					},
				},
			);

			dts_writer.update_file(id, types);

			return {
				code: s.toString(),
				map: s.generateMap(),
			};
		},
	};
}
