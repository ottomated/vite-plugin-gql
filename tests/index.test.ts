/// <reference types="@types/bun" />
import { test, expect } from 'bun:test';
import { generate_typescript } from '../src/codegen';
import { loadSchema } from '@graphql-tools/load';
import { format } from 'prettier';

const schema = await loadSchema(
	/* GraphQL */ `
		schema {
			query: QueryRoot
			mutation: Mutation
		}
		type QueryRoot {
			list: ListResult
			get(id: ID!): GetResult
		}
		type Mutation {
			create(input: CreateInput!): GetResult!
		}

		input CreateInput {
			name: String!
			numbers: [Int!]
		}

		interface Node {
			id: ID!
		}

		type GetResult implements Node {
			id: ID!
			name: String!
			numbers: [Int!]
		}

		type ListResult {
			items: [GetResult]
		}
	`,
	{
		loaders: [],
	},
);

const cases = [
	/* GraphQL */ `
		query {
			list {
				items {
					id
					name
					numbers
				}
			}
		}
	`,
	/* GraphQL */ `
		query ($id: ID!) {
			get(id: $id) {
				id
				name
				numbers
			}
		}
	`,
	/* GraphQL */ `
		mutation ($input: CreateInput!) {
			create(input: $input) {
				id
				name
			}
		}
	`,
];

test.each(cases)('gql %#', async (query) => {
	const res = generate_typescript(query, schema, undefined);
	const variables = await format('type In = ' + res.variables, {
		parser: 'typescript',
	});
	expect(variables).toMatchSnapshot();
	const return_type = await format('type Out = ' + res.return_type, {
		parser: 'typescript',
	});
	expect(return_type).toMatchSnapshot();
});
