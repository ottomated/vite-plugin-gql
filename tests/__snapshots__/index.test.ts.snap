// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`gql 0 1`] = `
"type In = null;
"
`;

exports[`gql 0 2`] = `
"type Out =
  | {
      list?:
        | {
            items?:
              | Array<
                  | {
                      id: string;
                      name: string;
                      numbers?: Array<number> | null | undefined;
                    }
                  | null
                  | undefined
                >
              | null
              | undefined;
          }
        | null
        | undefined;
    }
  | null
  | undefined;
"
`;

exports[`gql 1 1`] = `
"type In = { id: string };
"
`;

exports[`gql 1 2`] = `
"type Out =
  | {
      get?:
        | {
            id: string;
            name: string;
            numbers?: Array<number> | null | undefined;
          }
        | null
        | undefined;
    }
  | null
  | undefined;
"
`;

exports[`gql 2 1`] = `
"type In = {
  input: { name: string; numbers?: Array<number> | null | undefined };
};
"
`;

exports[`gql 2 2`] = `
"type Out = { create: { id: string; name: string } } | null | undefined;
"
`;
